README
                SystemC Class Library (Rel. 2.3.3)
                ==================================

This is the release of the SystemC 2.3.3 Class Library.
This release includes TLM 2.0.5 (Transaction Level Modeling) code,
as described in the IEEE 1666-2011 Language Reference Manual.

-------------------------------------------------------------------------------
                IMPORTANT

1. This is the release of SystemC 2.3.3. This release contains bug fixes
   for SystemC 2.3.2 (and earlier) as well as adding some minor features.

2. This release is supported on the following platform combinations for
   which it has been tested:

    o 64-bit Linux (x86_64)
      (RedHat Enterprise 6; SuSE Enterprise Linux 11; Debian 9)
      - GNU C++ compiler versions gcc-4.2.2 through gcc-8.2.0
      - Clang C++ compiler versions clang-3.4 through clang-6.0
      - Intel C++ Compiler (ICC 15.0.0)

    o 64-bit Linux (x86_64) with 32-bit compiler (--host=i686-linux-gnu)
      (SuSE Enterprise Linux 11)
      - GNU C++ compiler versions gcc-4.2.2 through gcc-8.2.0
      - Intel C++ Compiler (ICC 15.0.0)

    o 64-bit Linux (aarch64)
      (Ubuntu 16.04)
      - GNU C++ compiler version gcc-4.5.0

    o 64-bit Mac OS X (x86_64)
      (10.12 Sierra)
      - Apple LLVM version 8.0 (clang-800.0.42.1)
      - GNU C++ compiler (MacPorts) versions gcc-4.9.0, gcc-5.4.0, gcc-6.3.0

    o Microsoft Windows
      (Windows Server 2008, Windows 10 Enterprise)
      - Microsoft Visual Studio 2010 (10.0) (Win32 and x64)
      - Microsoft Visual Studio 2013 (12.0) (Win32 and x64)
      - Microsoft Visual Studio 2015 (14.0) (Win32 and x64)
      - Microsoft Visual Studio 2017 (14.1) (Win32 and x64)

      Warning: The fixed-point datatypes are not working as defined on
               MSVC 2015/2017 (x64) in Release mode.  See RELEASENOTES.

    o Windows Server 2008 (WoW64), Msys 1.0.17(0.48/3/2)
      - MinGW-w64 GNU C++ compiler version 4.9.2 (x86_64)


  This release has not yet been tested or is known not to work as expected
  on the following formerly supported platforms:

    o GNU C++ compiler versions prior to 4.2.x (all platforms)
    o FreeBSD 9.0 or later (x86, x86_64) with GNU C++ compiler
    o HP-UX 11.00 or later with GNU C++ or HP C++ compiler
    o Sun/Oracle Solaris, with GNU C++ or Sun/Solaris Studio compiler
    o Mac OS X prior to 10.12 (Sierra) and/or on the x86, PowerPC architectures
    o Microsoft Visual C++ versions prior to 10.0 (2010)
    o Cygwin 1.7 or later (x86, x86_64) with GNU C++ compiler
    o Msys/MinGW32 GNU C++ compiler


3. (In)compatibility note

  While it is possible to build SystemC as a shared library (both on
  ELF-based systems and well as a DLL on Windows), there is no explicitly
  defined binary interface across different

    - SystemC implementations (e.g. from multiple vendors)
    - SystemC library versions
    - Compilers or compiler versions
    - Operating systems or OS versions

  Some differences might be caught at link time, while other incompatibilities
  could just silently lead to undefined behavior during simulation.  Please
  carefully ensure a consistent build configuration, when integrating binary
  SystemC models built from separate sources.


-------------------------------------------------------------------------------

For details, see the separate RELEASENOTES file.


Licensing and Copyright

   See the separate LICENSE and NOTICE files to determine your rights
   and responsibilities for using SystemC.

User Documentation

   You can find documentation for this release in the docs directory.

Installation

   See the separate INSTALL file that provides system
   information and installation instructions.

Release Notes

   See the separate RELEASENOTES file that provides up to date
   information about this release of SystemC.

Additional information

   SystemC has a web site at

       http://www.accellera.org

   Discussion forum: http://forums.accellera.org/forum/9-systemc/

   You can post the bugs and suggestions of general interest to the forum.
   When reporting bugs please specify the following information
   (if applicable):

     1) SystemC version
     2) platform, compiler, flags
     3) description of the problem
     4) steps to reproduce the problem
     5) compile/runtime warnings and errors
     6) code sample, not more than 100 lines to demonstrate the problem

   Note: All bugs will only be tested against the latest publicly available
         version of the product.

   Note: All C++ compilers that SystemC supports have bugs of different
         degree of severity. We cannot fix those bugs. Please report them
         to the compiler vendor.


RELEASENOTES
                Release Notes for SystemC 2.3.3
                ===============================

        Andrew C. Goodrich,  Cadence Design Systems
        Philipp A. Hartmann, Intel Corporation

CONTENTS
========

  1) What's new in this release?

  2) Bug fixes and enhancements

  3) New features

  4) Incompatibilities with previous releases

  5) Initial support for C++11/14

  6) Experimental features

  7) Known problems

  8) Fixed-point library

  9) TLM Release Notes


1) What's new in this release?
==============================

This version of SystemC contains the "Proof of Concept" simulator
for the IEEE 1666-2011 SystemC standard. Please consult the IEEE Std
1666-2011 SystemC Language Reference Manual for details about the
current SystemC standard.

TLM-2.0 is merged into the main SystemC release since 2.3.0.
Please see section 9) below for more details related to TLM.

Compared to the 2.3.0 release, this release has the following new items:

  - New features, partially beyond the current IEEE 1666-2011 standard,
    see section 3.

  - Experimental features (some disabled by default), see section 5.
    Testing and feedback welcome via the Accellera SystemC forums
    at http://forums.accellera.org/forum/9-systemc/.

  - Bug fixes, see section 2.

  - Updated platform support, see the README.


2) Bug fixes and enhancements
=============================

For bug fixes added in 2.3.1 (or before) already, please refer to the previous
RELEASENOTES.

  - For fixes and enhancements of the embedded TLM-2.0 implementation,
    please see section 9.

  - Additional changes and cleanups leading to incompatibilities with
    earlier versions of SystemC are described in section 4.

Following is the list of bug fixes and enhancements for the 2.3.3 release:

  - sc_fxval: consider carry bit during addition/subtraction

  - sc_trace: fix overflow check for tracing signed C integer types

  - sc_trace: fix division by zero when closing an empty trace file

  - sc_trace_file::set_time_unit: report error for invalid timescale units

  - sc_elab_and_sim: guard against modification of argv from sc_main

  - sc_event: remove dynamic sensitivity of processes in destructor

  - Report error when calling wait/next_trigger on empty event lists

  - sc_vector_iter: avoid using std::iterator (deprecated in C++17)

  - Use proper bit in sc_signed_bitref::concat_set

  - Fix immediate triggering of reset during wait(n)

  - Detect calling wait(n) with n <= 0 in (clocked) threads

  - Fix memory corruption during simulation teardown when processes were
    spawned during end_of_elaboration/start_of_simulation

  - AArch64: fix stack corruption with optimizations disabled

  - sc_fifo_in/out: delete event finder objects

  - sc_port_base: free bind info in destructor

  - sc_host_mutex/semaphore: use C++11, if available

  - sc_start(): detect delta event notifications as activity

Following is the list of bug fixes and enhancements for the 2.3.2 release:

  - Adding support for modern compilers and related cleanups
    (MSVC 2015, recent GCC/Clang versions)

  - Remove conflict with external Boost library on Intel compiler

  - Fix bitvector element comparisons with char literals

    In 2.3.1, a new boolean conversion for sc_bitref_r had been
    added, breaking comparisons like bv[0] == '0'.  This extension
    has been reverted in 2.3.2 and replaced by an "explicit" boolean
    conversion instead (requires C++11, see section 5).

  - Improve sc_name_gen memory footprint

  - Add support for binding an sc_export during the
    before_end_of_elaboration() phase.

  - Detect instantiation of an sc_export outside of an sc_module

  - Detect invalid immediate notification of events during elaboration
    (report an error, instead of causing a crash in some cases)

  - Guard against self-assignment of sc_(and|or)_event_list objects

  - Fix incomplete time advance under the SC_RUN_TO_TIME starvation policy,
    causing incorrect sc_signal::event() values in some corner cases

  - Fix memory corruption when starting processes in asynchronous
    reset state (async_reset_signal_is asserted at startup)

  - Avoid accessing invalid memory in the sc_signal writer check
    in case conflicting writer processes have already terminated.

  - Correctly detect all conflicting signal writes under the
    SC_MANY_WRITERS writer policy.

  - Improved error checking in sc_fifo implementation

  - Minor enhancements of sc_vector-related classes
     - enable initialization from outside the current SystemC hierarchy
     - add stricter check of custom Creator compatibility
     - sc_vector_iter: enable comparison of compatible iterators

  - Avoid deleting the main fiber on Windows in some cases

  - Improve robustness of non-standard SC_WAIT macros to avoid
    giving wrong information when mixed with non-macro wait() calls

  - Fix timestamps when enabling tracing of delta cycles,
    improved tracing implementation

  - Catch unexpected calls to sc_start (e.g. from within a SystemC process)

  - Properly clean up (terminated) processes, releasing their resources
    once no longer referenced through a process handle

  - Add check for matching C++ standard selection (see section 5)

  - When using pthreads for SystemC thread processes, create them in
    the "detached" state to reduce resource pressure

  - In the default report handler, close the current and open a new logfile
    when calling sc_report_handler::set_log_file_name during simulation

  - Fix non-standard sc_simcontext::reset
     - properly handle sc_reset::reconcile_resets
     - delete elements in reverse order of creation

  - Cleanup systemc.h
     - Drop inclusion of 'cstring', 'sstream' by default (see section 4)
     - Avoid referencing deprecated 'std::gets' on C++11


3) New features
===============

Here is an overview of changes in 2.3.x compared to 2.3.0.

Note: These features partly add functionality beyond the current

src/sysc/kernel/sc_reset.h
/*****************************************************************************

  Licensed to Accellera Systems Initiative Inc. (Accellera) under one or
  more contributor license agreements.  See the NOTICE file distributed
  with this work for additional information regarding copyright ownership.
  Accellera licenses this file to you under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with the
  License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
  implied.  See the License for the specific language governing
  permissions and limitations under the License.

 *****************************************************************************/

/*****************************************************************************

  sc_reset.h -- Process reset support.

  Original Author: Andy Goodrich, Forte Design Systems, 17 June 2003

  CHANGE LOG AT THE END OF THE FILE
 *****************************************************************************/

#if !defined(sc_reset_h_INCLUDED)
#define sc_reset_h_INCLUDED

#include <ostream>
#include <vector>

#include "sysc/communication/sc_writer_policy.h"
#include "sysc/communication/sc_signal.h"
#include "sysc/kernel/sc_cmnhdr.h"
#include "sysc/kernel/sc_process.h"

#if defined(_MSC_VER) && !defined(SC_WIN_DLL_WARN)
#pragma warning(push)
#pragma warning(disable: 4251) // DLL import for std::vector
#endif

namespace sc_core {

// FORWARD CLASS REFERENCES:

template<typename DATA> class sc_signal_in_if;
template<typename DATA> class sc_in;
template<typename DATA> class sc_inout;
template<typename DATA> class sc_out;
class sc_reset; // IWYU pragma: keep

//==============================================================================
// CLASS sc_reset_target - RESET ENTRY FOR AN sc_process_b TARGET
//
// This class describes a reset condition associated with an sc_process_b
// instance. 
//==============================================================================
class SC_API sc_reset_target {
  public:
    bool          m_async;     // true asynchronous reset, false synchronous.
    bool          m_level;     // level for reset.
    sc_process_b* m_process_p; // process this reset entry is for.
};

inline std::ostream& operator << ( std::ostream& os, 
                                   const sc_reset_target& target )
{
    os << "[";
    os << target.m_async << ",";
    os << target.m_level << ",";
    os << target.m_process_p << ",";
    return os;
}


//==============================================================================
// sc_reset_finder - place holder class for a port reset signal until it is
//                   bound and an interface class is available. When the port
//                   has been bound the information in this class will be used
//                   to initialize its sc_reset object instance.
//==============================================================================
class SC_API sc_reset_finder {
    friend class sc_reset;
    friend class sc_simcontext;
public:
    sc_reset_finder( bool async, const sc_in<bool>* port_p, bool level,
                     sc_process_b* target_p);
    sc_reset_finder( bool async, const sc_inout<bool>* port_p, bool level,
                     sc_process_b* target_p);
    sc_reset_finder( bool async, const sc_out<bool>* port_p, bool level,
                     sc_process_b* target_p);

protected:
    bool                   m_async;     // True if asynchronous reset.
    bool                   m_level;     // Level for reset.
    sc_reset_finder*       m_next_p;    // Next reset finder in list.
    const sc_in<bool>*     m_in_p;      // Port for which reset is needed.
    const sc_inout<bool>*  m_inout_p;   // Port for which reset is needed.
    const sc_out<bool>*    m_out_p;     // Port for which reset is needed.
    sc_process_b*          m_target_p;  // Process to reset.

private: // disabled
    sc_reset_finder( const sc_reset_finder& );
    const sc_reset_finder& operator = ( const sc_reset_finder& );
};

//==============================================================================
// CLASS sc_reset - RESET INFORMATION FOR A RESET SIGNAL
//
// See the top of sc_reset.cpp for an explaination of how the reset mechanism
// is implemented.
//==============================================================================
class SC_API sc_reset {
    friend class sc_cthread_process;
    friend class sc_method_process; 
    friend class sc_module; 
    friend class sc_process_b;
    friend class sc_signal<bool, SC_ONE_WRITER>;
    friend class sc_signal<bool, SC_MANY_WRITERS>;
    friend class sc_signal<bool, SC_UNCHECKED_WRITERS>;
    friend class sc_simcontext;
    template<typename SOURCE> friend class sc_spawn_reset;
    friend class sc_thread_process;

  protected:
    static void reconcile_resets(sc_reset_finder* reset_finder_q);
    static void 
        reset_signal_is(bool async, const sc_signal_in_if<bool>& iface, 
                        bool level);
    static void 
        reset_signal_is( bool async, const sc_in<bool>& port, bool level);
    static void 
        reset_signal_is( bool async, const sc_inout<bool>& port, bool level);
    static void 
        reset_signal_is( bool async, const sc_out<bool>& port, bool level);

  protected:
    sc_reset( const sc_signal_in_if<bool>* iface_p ) :
        m_iface_p(iface_p), m_targets() {}
    void notify_processes();
    void remove_process( sc_process_b* );

  protected:
    const sc_signal_in_if<bool>*  m_iface_p;  // Interface to read.
    std::vector<sc_reset_target>  m_targets;  // List of processes to reset.

  private: // disabled
    sc_reset( const sc_reset& );
    const sc_reset& operator = ( const sc_reset& );
};

// $Log: sc_reset.h,v $
// Revision 1.11  2011/08/26 20:46:10  acg
//  Andy Goodrich: moved the modification log to the end of the file to
//  eliminate source line number skew when check-ins are done.
//
// Revision 1.10  2011/08/24 22:05:51  acg
//  Torsten Maehne: initialization changes to remove warnings.
//
// Revision 1.9  2011/04/08 22:38:30  acg
//  Andy Goodrich: added comment pointing to the description of how the
//  reset mechanism works that is in sc_reset.cpp.
//
// Revision 1.8  2011/02/18 20:27:14  acg
//  Andy Goodrich: Updated Copyrights.
//
// Revision 1.7  2011/02/13 21:47:37  acg
//  Andy Goodrich: update copyright notice.
//
// Revision 1.6  2011/01/06 18:00:32  acg
//  Andy Goodrich: Removed commented out code.
//
// Revision 1.5  2010/12/07 20:09:14  acg
// Andy Goodrich: removed sc_signal signatures since already have sc_signal_in_if signatures.
//
// Revision 1.4  2010/11/20 17:10:57  acg
//  Andy Goodrich: reset processing changes for new IEEE 1666 standard.
//
// Revision 1.3  2009/05/22 16:06:29  acg
//  Andy Goodrich: process control updates.
//
// Revision 1.2  2008/05/22 17:06:26  acg
//  Andy Goodrich: updated copyright notice to include 2008.
//
// Revision 1.1.1.1  2006/12/15 20:20:05  acg
// SystemC 2.3
//
// Revision 1.6  2006/12/02 20:58:19  acg
//  Andy Goodrich: updates from 2.2 for IEEE 1666 support.
//
// Revision 1.4  2006/04/11 23:13:21  acg
//   Andy Goodrich: Changes for reduced reset support that only includes
//   sc_cthread, but has preliminary hooks for expanding to method and thread
//   processes also.
//
// Revision 1.3  2006/01/13 18:44:30  acg
// Added $Log to record CVS changes into the source.

} // namespace sc_core

#if defined(_MSC_VER) && !defined(SC_WIN_DLL_WARN)
#pragma warning(pop)
#endif

#endif // !defined(sc_reset_h_INCLUDED)
